### PR TITLE
Discovery SPI minor changes.

### PR DESCRIPTION
**Pull request by kobalski**
Adds fail fast to JoinConfig for Discovery SPI when enabled with TCP/IP discovery.
Adds a property to enable use of public IP addresses in DiscoveryJoiner.

### CODE REVIEW

**Comment by noctarius** [comment]
It is perfectly legal to have predefined IPs + discovery, this is for mixed private+public clouds. You're right about AWS though.

**Comment by kobalski** [comment]
Ok, removed restriction on having predefined IPs + discovery together. Thanks for the warning.

**Comment by kobalski** [comment]
verify

File: hazelcast/src/main/java/com/hazelcast/config/JoinConfig.java [hunk 1]
```diff
@@ -118,16 +118,27 @@ public void verify() {
...
             throw new InvalidConfigurationException("Multicast and AWS join can't be enabled at the same time");
         }
+        verifyDiscoveryProviderConfig();
+    }
 
+    /**
+     * Verifies this JoinConfig is valid. When Discovery SPI enabled other discovery
+     * methods should be disabled
+     *
+     * @throws IllegalStateException when the join config is not valid.
```
Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/config/JoinConfig.java]: `@throws IllegalStateException when the join config is not valid.` <-- that is not correct.

**Comment by mesutcelik** [comment]
:+1:

**Comment by kobalski** [comment]
verify

**Comment by kobalski** [comment]
verify

**Comment by bilalyasar** [comment]
LGTM :+1: , also needs some docs change.

**Comment by kobalski** [comment]
verify

**Pull request merged**